src/klein/resource.py

```diff
@@ -278,3 +278,97 @@ def processing_failed(failure, error_handlers):
         d.addErrback(processing_failed, self._app._error_handlers)
         d.addCallback(write_response).addErrback(log.err, _why="Unhandled Error writing response")
         return server.NOT_DONE_YET
+
+class Arguments(object):
+    """
+    A wrapper around the C{dict} type that gives easy access to the
+    C{bytes} keys set by Twisted.
+    """
+
+    def __init__(self, args=None):
+        self.arguments = {} if not args else args
+
+    def __contains__(self, item):
+        """
+        Check if the item is a key in the arguments.
+
+        @returns: C{bool}
+        """
+        item = ensure_utf8_bytes(item)
+        return item in self.arguments
+
+    def __delitem__(self, key):
+        del self.arguments[key]
+
+    def __getattr__(self, dict_attr):
+        """
+        Access attributes and functions from the underlaying dict.
+        """
+        return getattr(self.arguments, dict_attr)
+
+    def __getitem__(self, key):
+        """
+        Return the value from the arguments. Tries to be smart and gets
+        a single value if only 1 item exists in the list or entire list.
+
+        @params key: C{bytes}, C{unicode}, or C{str}
+        """
+        key = ensure_utf8_bytes(key)
+        # return 1 item or all items
+        result = self.arguments.get(key, [None])
+        if len(result) == 1:
+            return result[-1]
+        return result
+
+    def __len__(self):
+        return len(self.arguments)
+
+    def __setitem__(self, key, value):
+        """
+        Set a value in the dict. The key will be converted into C{bytes}
+        and can be accessed using the C{bytes}, C{unicode}, or C{str}.
+
+        @param key: C{bytes}, C{unicode}, or C{str}
+        """
+        key = ensure_utf8_bytes(key)
+        self.arguments.setdefault(key, []).append(value)
+
+    def __iter__(self):
+        return iter(self.arguments)
+
+    def get(self, key, default=None):
+        """
+        Retrieve the value from the dict. Unlike ``__getitem__``,
+        this function will not return a single value if there's
+        only 1 in the list. Entire list, None or default value will
+        be returned.
+
+        @param key: C{bytes}, C{unicode}, or C{str}
+        @param default: User specified default value.
+        """
+        key = ensure_utf8_bytes(key)
+        return self.arguments.get(key, default)
+
+class KleinHTTPRequest(server.Request):
+    _arguments = None
+
+    @property
+    def arguments(self):
```

Comment:
See the example in my [comment](https://github.com/twisted/klein/pull/158#issuecomment-270293713)